### ISSUE TITLE
[no-unused-expressions] no-unused-expressions works not as expected with optional chaining and condition

### ISSUE THREAD

**Issue by kkatsuba** [typescript-eslint/typescript-eslint#1764]
**Repro**

```JSON
{
  "rules": {
    "no-unused-expressions": 0,
    "@typescript-eslint/no-unused-expressions": [2, {
      "allowShortCircuit": true
    }]
  }
}
```

```TS
const expr: any = false;
const v: { [key: string]: any } = {};
v.somestaff?.callMe(); // Works fine 👍
expr && v.somestaff?.callMe(); // Problem 😭
```

**Expected Result**
No error for line

**Actual Result**
![image](https://user-images.githubusercontent.com/17414540/76959728-3ca24880-692b-11ea-8c17-8a4b5cda8a15.png)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.23.0` |
| `@typescript-eslint/parser`        | `2.23.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `11.15.0` |
| `npm`                              | `6.7.0` |

**Comment by InsOpDe** [comment]
Similar here:
```
dispatch?.()
```
throws the same warning
